hkmc2/shared/src/main/scala/hkmc2/codegen/Lifter.scala

```diff
@@ -0,0 +1,1504 @@
+    val captureSym = FlowSymbol("capture")
+    val captureCtx = ctx
+      .addLocalCaptureSyms(varsMap) // how to access locals via. the capture class from now on
+      .addCapturePath(f.sym, captureSym.asPath) // the path to this function's capture
+      .addLocalPaths((thisVars.vars.toSet -- thisVars.reqCapture).map(s => s -> s).toMap)
+      .addLocalPaths(modPaths)
+      .addIgnoredBmsPaths(ignored.map(d => d.sym -> d.sym).toMap)
+    val nestedCtx = captureCtx.addFnLocals(captureCtx.usedLocals(f.sym))
+
+    // lift out the nested defns
+    val nestedLifted = included.map(liftOutDefnCont(f, _, nestedCtx.flushModules))
+    val ignoredLifted = ignored.map(liftOutDefnCont(f, _, nestedCtx))
+    val ignoredExtra = ignoredLifted.flatMap(_.extraDefns)
+    val newDefns = ignoredExtra ++ nestedLifted.flatMap:
+      case Lifted(liftedDefn, extraDefns) => liftedDefn :: extraDefns
+      
+    val ignoredRewrite = ignoredLifted.map: lifted =>
+        lifted.liftedDefn.sym -> lifted.liftedDefn
+      .toMap
+
+    val transformed = rewriteBlk(blk, N, captureCtx.addReplDefns(ignoredRewrite))
+
+    if thisVars.reqCapture.size == 0 then
+      Lifted(FunDefn(f.owner, f.sym, f.params, transformed), newDefns)
+    else
+      // move the function's parameters to the capture
+      val paramsSet = f.params.flatMap(_.paramSyms)
+      val paramsList = varsList.map: s =>
+        if paramsSet.contains(s) then s.asPath else Value.Lit(Tree.UnitLit(true))
+      // moved when the capture is instantiated
+      val bod = blockBuilder
+        .assign(captureSym, Instantiate(captureCls.sym.asPath, paramsList))
+        .rest(transformed)
+      Lifted(FunDefn(f.owner, f.sym, f.params, bod), captureCls :: newDefns)
+
+  end liftDefnsInFn
+
+  def desugarLambdas(b: Block) =
+    def rewriteOneBlk(b: Block) =
+      var lambdasList: List[(BlockMemberSymbol, Value.Lam)] = Nil
+      val lambdaRewriter = new BlockTransformerNoRec(SymbolSubst()):
+        override def applyValue(v: Value): Value = v match
+          case lam: Value.Lam => 
+            val sym = BlockMemberSymbol("lambda", Nil)
+            lambdasList ::= (sym -> super.applyLam(lam))
+            Value.Ref(sym)
+          case _ => super.applyValue(v)
+      val blk = lambdaRewriter.applyBlock(b)
+      (blk, lambdasList)
+
+    val transformer = new BlockTransformer(SymbolSubst()):
+      override def applyBlock(b: Block): Block =
+        val (newBlk, lambdasList) = rewriteOneBlk(b)
+        val lambdaDefns = lambdasList.map:
+          case (sym, Value.Lam(params, body)) =>
+            FunDefn(None, sym, params :: Nil, body)
+        val ret = lambdaDefns.foldLeft(newBlk):
+          case (acc, defn) => Define(defn, acc)
+        super.applyBlock(ret)
+    transformer.applyBlock(b)
+
+  // top-level
+  def transform(b: Block) =
+    // this is already done once in the lowering, but the handler lowering adds lambdas currently
```

Comment:
Then shouldn't the code to do this be written only once and shared?

Reply:
In fact I just noticed the exact same code is also added in `Lowering`, which looks like a mistake.